### DISCUSSION TITLE
Fetching entities not by `Entity`?

### DISCUSSION THREAD

**Discussion by ghost** [bevyengine/bevy#3543] [Q&A] [answered]
Please keep in mind I'm not an experienced game developer. Here's a problem I hit with Bevy:

I've got entities that can reference each other using `Entity`. But `Entity` is only known after it gets spawned and can change as well if it's deleted/respawned, right? I therefore assumed that I need a stronger, persistent id for my entities, a way for entities to refer to each other even without being spawned. But then, how would I find entities by that persistent id - wouldn't iterating queries in search of a match be a waste of resources as opposed to using `HashMap` and being able to find my entity in constant time?

If I'm not using Bevy, I would have something like this:

```rust
struct World {
  rooms: HashMap<RoomId, Room>,
}

struct RoomId(String);

struct Room {
  persistent_id: RoomId,
  exits: Vec<RoomId>,
}
```

And then, as a dumb example, whenever I have a `Room` and want to resolve one of its exits into another `Room` by the `RoomId`, I'd just `world::rooms.get(room_id)`. But I don't see how I can do that in Bevy?

**Comment by CGMossa** [comment]
I'd say take `Entity` as being your persistent identifier (id), because that's what entities are.
I don't understand your use-case exactly, but generally speaking, you would want to 
spawn entities, and maybe add a marker type to it, e.g. `struct Room;`, so that when you query at any
given time, you'd retrieve all entities, that happens to be a `Room`, and then go from there.

Can I suggest joining us over at the discord, and especially at the `#help`-channel? I have many questions like this over there, and people have responded very promptly, and also guided me through some of this.
You can search for these conversations, or just start a thread there.

**Comment by DJMcNab** [comment]
To add further to this, you can spawn an entity with no components, then add components at runtime.

**Comment by cart** [comment, marked as answer]
The Entity id should be the default and preferred choice for referencing entities at runtime, because you get it for free and it is the cheapest and most direct way to access an entity's components. However if you need a "persistent" and a stable id (ex: networking, some types of scenes persisted to disk), a `HashMap<MyId, Entity>` resource is a perfectly reasonable solution to the problem.
